File: hdar/archives/hdar_3/hdar/cpp/src/capabilities.cpp
#include "hdar/capabilities.hpp"
#include <algorithm>
#include <sstream>

namespace hdar {

JsonValue Capability::to_json() const {
    JsonValue v = JsonValue::object();
    v["name"] = JsonValue::string(name);
    v["scope"] = JsonValue::string(scope);
    v["granted"] = JsonValue::boolean(granted);
    JsonValue cons = JsonValue::object();
    for (const auto& [k, val] : constraints)
        cons[k] = JsonValue::string(val);
    v["constraints"] = std::move(cons);
    return v;
}

Capability Capability::from_json(const JsonValue& v) {
    Capability c;
    c.name = v.get("name").string_val;
    c.scope = v.get("scope").string_val;
    c.granted = v.get("granted").bool_val;
    const auto& cons = v.get("constraints");
    if (cons.type == JsonValue::Type::Object) {
        for (const auto& [k, val] : cons.object_val)
            c.constraints[k] = val.string_val;
    }
    return c;
}

bool is_scope_broader(const std::string& src, const std::string& dst) {
    if (dst == "*" && src != "*") return true;
    if (src == "/workspace" && dst == "/") return true;
    if (src == "/workspace" && dst == "*") return true;
    if (dst == src) return false;
    if (dst.size() > src.size() && dst.substr(0, src.size() + 1) == src + "/")
        return false; // subdirectory is narrower
    if (src.size() > dst.size() && src.substr(0, dst.size() + 1) == dst + "/")
        return false; // src is subdirectory of dst, dst is broader — but this means dst IS broader
    // Everything else is broader
    return true;
}

bool is_budget_higher(const std::string& src, const std::string& dst) {
    auto parse = [](const std::string& s) -> double {
        std::string clean;
        for (char c : s) {
            if (c != '$' && c != ',') clean += c;
        }
        try { return std::stod(clean); }
        catch (...) { return 0.0; }
    };
    return parse(dst) > parse(src);
}

CapabilityCompiler::CapabilityCompiler() {}

std::pair<std::vector<Capability>, std::vector<std::string>>
CapabilityCompiler::compile(const std::vector<Capability>& source_caps,
                            const std::map<std::string, std::string>& policy) {
    std::vector<Capability> dest_caps;
    std::vector<std::string> rejections;

    for (const auto& src : source_caps) {
        if (!src.granted) continue;

        std::pair<std::optional<Capability>, std::string> result{{}, ""};

        if (src.name == "filesystem.read" || src.name == "filesystem.write")
            result = map_filesystem(src, policy);
        else if (src.name == "network.egress")
            result = map_network(src, policy);
        else if (src.name == "budget.spend")
            result = map_budget(src, policy);
        else if (src.name == "deploy")
            result = map_deploy(src, policy);
        else if (src.name == "shell.exec")
            result = map_shell(src, policy);
        else {
            rejections.push_back("unknown capability '" + src.name + "' — denied by default");
            continue;
        }

        if (result.first)
            dest_caps.push_back(*result.first);
        else
            rejections.push_back(result.second);
    }

    return {dest_caps, rejections};
}

std::pair<std::optional<Capability>, std::string>
CapabilityCompiler::map_filesystem(const Capability& src,
                                    const std::map<std::string, std::string>& policy) {
    auto it = policy.find("filesystem.root");
    std::string dst_scope = (it != policy.end()) ? it->second : "/workspace";
    if (is_scope_broader(src.scope, dst_scope))
        return {std::nullopt, "filesystem scope '" + dst_scope + "' is broader than source '" +
                src.scope + "' — capability broadening rejected"};
    return {Capability{src.name, dst_scope, true, src.constraints}, ""};
}

std::pair<std::optional<Capability>, std::string>
CapabilityCompiler::map_network(const Capability& src,
                                 const std::map<std::string, std::string>& policy) {
    auto it = policy.find("network.allowlist");
    std::string allowlist = (it != policy.end()) ? it->second : "";
    if (!allowlist.empty() && src.scope != "*") {
        std::vector<std::string> allowed;
        std::stringstream ss(allowlist);
        std::string token;
        while (std::getline(ss, token, ',')) {
            // trim
            while (!token.empty() && token.front() == ' ') token.erase(0, 1);
            while (!token.empty() && token.back() == ' ') token.pop_back();
            allowed.push_back(token);
        }
        bool found = false;
        for (const auto& a : allowed) {
            if (a == src.scope || a == "*") { found = true; break; }
        }
        if (!found)
            return {std::nullopt, "network scope '" + src.scope +
                    "' not in destination allowlist [" + allowlist + "] — capability denied"};
    }
    if (allowlist.empty() && src.scope == "*")
        return {std::nullopt, "wildcard network not allowed in destination — denied"};

    Capability c{src.name, src.scope, true, {}};
    if (!allowlist.empty()) c.constraints["allowlist"] = allowlist;
    return {c, ""};
}

std::pair<std::optional<Capability>, std::string>
CapabilityCompiler::map_budget(const Capability& src,
                                const std::map<std::string, std::string>& policy) {
    auto it = policy.find("budget.max");
    std::string dst_budget = (it != policy.end()) ? it->second : src.scope;
    if (is_budget_higher(src.scope, dst_budget))
        return {std::nullopt, "budget '" + dst_budget + "' exceeds source '" +
                src.scope + "' — capability broadening rejected"};
    return {Capability{src.name, dst_budget, true, {}}, ""};
}

std::pair<std::optional<Capability>, std::string>
CapabilityCompiler::map_deploy(const Capability& src,
                                const std::map<std::string, std::string>& policy) {
    auto it = policy.find("deploy.allowed");
    if (it == policy.end() || it->second != "true")
        return {std::nullopt, "deploy not allowed in destination policy — denied"};
    return {Capability{src.name, src.scope, true, {}}, ""};
}

std::pair<std::optional<Capability>, std::string>
CapabilityCompiler::map_shell(const Capability& src,
                               const std::map<std::string, std::string>& policy) {
    auto it = policy.find("shell.allowed");
    if (it == policy.end() || it->second != "true")
        return {std::nullopt, "shell not allowed in destination policy — denied"};
    return {Capability{src.name, src.scope, true, {}}, ""};
}

std::pair<bool, std::vector<std::string>>
CapabilityCompiler::verify_non_expansion(const std::vector<Capability>& source_caps,
                                          const std::vector<Capability>& dest_caps) const {
    std::vector<std::string> violations;
    std::map<std::string, const Capability*> src_map;
    for (const auto& c : source_caps)
        if (c.granted) src_map[c.name] = &c;

    for (const auto& dst : dest_caps) {
        if (!dst.granted) continue;
        auto it = src_map.find(dst.name);
        if (it == src_map.end()) {
            violations.push_back("destination grants '" + dst.name +
                                 "' not present in source — expansion");
            continue;
        }
        const auto& src = *it->second;
        if (dst.name.rfind("filesystem", 0) == 0) {
            if (is_scope_broader(src.scope, dst.scope))
                violations.push_back("filesystem scope expanded: '" +
                                     src.scope + "' → '" + dst.scope + "'");
        } else if (dst.name == "budget.spend") {
            if (is_budget_higher(src.scope, dst.scope))
                violations.push_back("budget expanded: '" +
                                     src.scope + "' → '" + dst.scope + "'");
        }
    }

    return {violations.empty(), violations};
}

} // namespace hdar

File: hdar/capsule/capabilities.py
"""Capability continuity — authority mapping between providers.

The central invariant: migration may preserve or reduce authority,
but it may never silently expand it.

Capabilities are typed, deny-by-default. Each capability has:
  - name: e.g. "filesystem.write", "network.egress", "budget.spend"
  - scope: e.g. "/workspace", "api.example.com", "$5"
  - granted: bool

The compiler maps source capabilities to destination capabilities.
Any mapping that would broaden authority is rejected.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple


@dataclass
class Capability:
    """A single typed capability grant."""
    name: str          # e.g. "filesystem.write"
    scope: str         # e.g. "/workspace" or "api.example.com"
    granted: bool = True
    constraints: Dict[str, str] = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "scope": self.scope,
            "granted": self.granted,
            "constraints": self.constraints,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "Capability":
        return cls(
            name=d["name"],
            scope=d["scope"],
            granted=d.get("granted", True),
            constraints=d.get("constraints", {}),
        )


# ─── Scope comparison ─────────────────────────────────────

def is_scope_broader(src_scope: str, dst_scope: str) -> bool:
    """Returns True if dst_scope is broader than src_scope."""
    if dst_scope == "*" and src_scope != "*":
        return True
    if src_scope == "/workspace" and dst_scope == "/":
        return True
    if src_scope == "/workspace" and dst_scope == "*":
        return True
    # Exact match or narrower is fine
    if dst_scope == src_scope:
        return False
    # Subdirectory is narrower
    if dst_scope.startswith(src_scope + "/"):
        return False
    # Everything else is broader
    if not src_scope.startswith(dst_scope + "/"):
        return True
    return False


def is_budget_higher(src: str, dst: str) -> bool:
    """Compare budget scopes. Returns True if dst allows more than src."""
    def parse(s: str) -> float:
        s = s.replace("$", "").replace(",", "").strip()
        try:
            return float(s)
        except ValueError:
            return 0.0
    return parse(dst) > parse(src)


# ─── Capability compiler ─────────────────────────────────

class CapabilityCompiler:
    """Maps source capabilities to destination provider capabilities.

    Enforces the invariant: authority may be preserved or reduced,
    but never silently increased.
    """

    def __init__(self):
        self._rules: Dict[str, callable] = {}
        self._register_defaults()

    def _register_defaults(self):
        """Register default mapping rules for known capability types."""
        self._rules["filesystem.read"] = self._map_filesystem
        self._rules["filesystem.write"] = self._map_filesystem
        self._rules["network.egress"] = self._map_network
        self._rules["budget.spend"] = self._map_budget
        self._rules["deploy"] = self._map_deploy
        self._rules["shell.exec"] = self._map_shell

    def compile(
        self,
        source_caps: List[Capability],
        destination_policy: Dict[str, str],
    ) -> Tuple[List[Capability], List[str]]:
        """Map source capabilities to destination.

        Returns (destination_caps, rejections).
        Each rejection explains why a capability was denied.
        """
        dest_caps: List[Capability] = []
        rejections: List[str] = []

        for src_cap in source_caps:
            if not src_cap.granted:
                continue

            rule = self._rules.get(src_cap.name)
            if rule is None:
                # Unknown capability — deny by default
                rejections.append(
                    f"unknown capability '{src_cap.name}' — denied by default"
                )
                continue

            dst_cap, reason = rule(src_cap, destination_policy)
            if dst_cap is not None:
                dest_caps.append(dst_cap)
            else:
                rejections.append(reason)

        return dest_caps, rejections

    def _map_filesystem(self, src: Capability, policy: Dict) -> Tuple[Optional[Capability], str]:
        dst_scope = policy.get("filesystem.root", "/workspace")
        if is_scope_broader(src.scope, dst_scope):
            return None, (
                f"filesystem scope '{dst_scope}' is broader than source '{src.scope}' "
                f"— capability broadening rejected"
            )
        return Capability(
            name=src.name,
            scope=dst_scope,
            granted=True,
            constraints=src.constraints,
        ), ""

    def _map_network(self, src: Capability, policy: Dict) -> Tuple[Optional[Capability], str]:
        allowlist = policy.get("network.allowlist", "")
        if allowlist and src.scope != "*":
            # Check if source scope is in the allowlist
            allowed = [a.strip() for a in allowlist.split(",")]
            if src.scope not in allowed and "*" not in allowed:
                return None, (
                    f"network scope '{src.scope}' not in destination allowlist "
                    f"[{allowlist}] — capability denied"
                )
        if allowlist == "" and src.scope == "*":
            return None, "wildcard network not allowed in destination — denied"
        return Capability(
            name=src.name,
            scope=src.scope,
            granted=True,
            constraints={"allowlist": allowlist} if allowlist else {},
        ), ""

    def _map_budget(self, src: Capability, policy: Dict) -> Tuple[Optional[Capability], str]:
        dst_budget = policy.get("budget.max", src.scope)
        if is_budget_higher(src.scope, dst_budget):
            return None, (
                f"budget '{dst_budget}' exceeds source '{src.scope}' "
                f"— capability broadening rejected"
            )
        return Capability(
            name=src.name,
            scope=dst_budget,
            granted=True,
        ), ""

    def _map_deploy(self, src: Capability, policy: Dict) -> Tuple[Optional[Capability], str]:
        deploy_allowed = policy.get("deploy.allowed", "false")
        if deploy_allowed != "true":
            return None, "deploy not allowed in destination policy — denied"
        return Capability(
            name=src.name,
            scope=src.scope,
            granted=True,
        ), ""

    def _map_shell(self, src: Capability, policy: Dict) -> Tuple[Optional[Capability], str]:
        shell_allowed = policy.get("shell.allowed", "false")
        if shell_allowed != "true":
            return None, "shell not allowed in destination policy — denied"
        return Capability(
            name=src.name,
            scope=src.scope,
            granted=True,
        ), ""

    def verify_non_expansion(
        self,
        source_caps: List[Capability],
        dest_caps: List[Capability],
    ) -> Tuple[bool, List[str]]:
        """Verify that destination capabilities do not expand source."""
        violations: List[str] = []
        src_map = {c.name: c for c in source_caps if c.granted}

        for dst in dest_caps:
            if not dst.granted:
                continue
            src = src_map.get(dst.name)
            if src is None:
                violations.append(
                    f"destination grants '{dst.name}' not present in source — expansion"
                )
                continue
            if dst.name.startswith("filesystem"):
                if is_scope_broader(src.scope, dst.scope):
                    violations.append(
                        f"filesystem scope expanded: '{src.scope}' → '{dst.scope}'"
                    )
            elif dst.name == "budget.spend":
                if is_budget_higher(src.scope, dst.scope):
                    violations.append(
                        f"budget expanded: '{src.scope}' → '{dst.scope}'"
                    )

        return len(violations) == 0, violations
